### PR TITLE
MFEM AWS tutorial

### PR DESCRIPTION
**Pull request by tzanio**
Pages for the MFEM tutorial as part of the [RADIUSS AWS Tutorial Series](https://software.llnl.gov/radiuss/event/2022/07/07/radiuss-on-aws/).

Screen Shot 2022-07-27 at 8 55 14 PM

### CODE REVIEW

**Comment by pazner** [comment]
```mermaid
  graph TD;
      A[Getting Started]-->B[Finite Element Basics];
      B-->C[Tour of MFEM Examples];
      B-->D[Meshing and Visualization];
      B-->E[Solvers and Scalability];
      C-->G[Further Steps];
      D-->G;
      E-->G;
```

File: src/tutorial/fem.md [hunk 1]
```diff
@@ -55,12 +55,12 @@ $$
...
+functions are typically piecewise-polynomial functions on a given computational mesh,
 which are only non-zero on small portions of the mesh.
 
 <img src="../img/fem1.png" width="300">
 
-With finite elements, the mesh can be totally unstructured, curved and
+With finite elements, the mesh can be totally unstructured, curved, and
 non-conforming:
 
 <img src="../img/fem2.png" width="300">
```
Comment (hauten) [inline src/tutorial/fem.md:66]: fem2.png is broken for me locally, even though the image file is in the correct directory.

File: src/tutorial/fem.md [hunk 2]
```diff
@@ -113,20 +113,21 @@ To learn more, you can visit MFEM's <a href="../../fem/">Finite Element Method</
...
+MFEM's Example 1 implements the above simple FEM for the Poisson problem
 in the source file [examples/ex1.cpp](https://github.com/mfem/mfem/blob/master/examples/ex1.cpp).
 We set $f=1$ in (1) and enforce homogeneous Dirichlet boundary conditions on the whole boundary.
 
-Below we highlight selected portion of the example code and connect them with
+Below we highlight selected portions of the example code and connect them with
 the description in the previous section. You can follow along by browsing
-`ex1.cpp` in your VS Code browser window.
+`ex1.cpp` in your VS Code browser window. In the settings of this tutorial, the
+visualization will automatically update in the GLVis browser window.
```
Comment (hauten) [inline src/tutorial/fem.md:123]: I moved this statement b/c it makes me think I'm supposed to see a change in the example viz, but at that point I haven't entered any new commands.

**Comment by hauten** [comment]
I inserted some comments in https://github.com/mfem/web/pull/176/commits/6cca4651caa4b1d1baeae502da03a03d3ce6a035. Also, line 383 of `fem.md`: fem4.png doesn't appear to show the results any of the sample commands in lines 378-380. Perhaps add it to the list of samples?

File: src/tutorial/examples.md [hunk 3]
```diff
@@ -27,17 +27,25 @@ and <a href="../fem"><i class="fa fa-book"></i> Finite Element Basics</a> pages
...
     - Make sure that the MFEM source code is already compiled
     - Enter the `examples` directory: `cd ~/mfem/examples` 
-    - Compile the serial version of `example ?` with `make ex?` or the parallel version with `make ex?p`. 
+
+<div class="panel panel-info" style="width:92%; margin-left: auto; margin-right: auto;">
+<div class="panel-heading">
+<h3 class="panel-title"><i class="fa fa-info-circle"></i>&nbsp; Note</h3>
+</div>
+<div class="panel-body">
+Remember to compile each numbered example before executing its sample runs: <code>make ex?</code> for the serial version or <code>make ex?p</code> for the parallel version. You can build multiple examples in the same command: <code>make ex3 ex4 ex3p ex4p</code>
```
Comment (hauten) [inline src/tutorial/examples.md:41]: I added this b/c I had a lot of errors and couldn't run anything after Example 1 until I got to `solvers.md` and realized I had to build the others.

File: src/tutorial/examples.md [hunk 4]
```diff
@@ -27,17 +27,25 @@ and <a href="../fem"><i class="fa fa-book"></i> Finite Element Basics</a> pages
...
+</div>
 
 <img class="floatright" src="../../img/examples/ex1.png", width="200"/>
 
 - **Example 1** ([ex1.cpp](https://github.com/mfem/mfem/blob/master/examples/ex1.cpp) and
-     [ex1p.cpp](https://github.com/mfem/mfem/blob/master/examples/ex1p.cpp)) solves a simple Poisson problem using a scalar $H^1$ space. More specificaly it solves the problem $$-\Delta u = 1$$ with homogeneous Dirichlet boundary conditions.
+     [ex1p.cpp](https://github.com/mfem/mfem/blob/master/examples/ex1p.cpp)) solves a simple Poisson problem using a scalar $H^1$ space. More specificaly, it solves the problem $$-\Delta u = 1$$ with homogeneous Dirichlet boundary conditions.
 
     Try the following sample runs:
     <code class="language-java"><pre>
```
Comment (hauten) [inline src/tutorial/examples.md:51]: For me, none of these samples results in a viz that looks like ex1.png. `fichera.mesh` looked more like ex3 but all blue. OK if the figures don't exactly correspond to the sample runs?

File: src/tutorial/examples.md [hunk 5]
```diff
@@ -27,17 +27,25 @@ and <a href="../fem"><i class="fa fa-book"></i> Finite Element Basics</a> pages
...
+</div>
+</div>
 
 <img class="floatright" src="../../img/examples/ex1.png", width="200"/>
 
 - **Example 1** ([ex1.cpp](https://github.com/mfem/mfem/blob/master/examples/ex1.cpp) and
-     [ex1p.cpp](https://github.com/mfem/mfem/blob/master/examples/ex1p.cpp)) solves a simple Poisson problem using a scalar $H^1$ space. More specificaly it solves the problem $$-\Delta u = 1$$ with homogeneous Dirichlet boundary conditions.
+     [ex1p.cpp](https://github.com/mfem/mfem/blob/master/examples/ex1p.cpp)) solves a simple Poisson problem using a scalar $H^1$ space. More specificaly, it solves the problem $$-\Delta u = 1$$ with homogeneous Dirichlet boundary conditions.
 
     Try the following sample runs:
```
Comment (hauten) [inline src/tutorial/examples.md:50]: In all files that contain `<code class="language-console"><pre>`, a blank space styled as code appears at the end of the line above (here it's after "Try the following sample runs:"). I couldn't get it to go away. Maybe that's fine.

File: src/tutorial/solvers.md [hunk 6]
```diff
@@ -152,7 +152,7 @@ each of these examples in more details below.
 
 - Try running `./ex2p` to run a 2D elasticity problem.
 
-- As in example 1, the linear system is solved using _algebraic multigrid_.
+- As in Example 1, the linear system is solved using AMG.
 
 - For this example, two types of AMG solvers can be used:
```
Comment (hauten) [inline src/tutorial/solvers.md:158]: GitHub won't let me comment exactly in these places:
Lines 164-167: Where in `ex2p` do I add the `-elast` flag or make the `--order` argument? Or do I change something in the command line?
Lines 230-238: Same as the question above, where do I make these changes? I would like to see additional sample run(s) or lines of code that I'm supposed to change as in 198-199 and 284-185.

File: src/tutorial/meshvis.md [hunk 7]
```diff
@@ -30,15 +30,13 @@ and <a href="../fem"><i class="fa fa-book"></i> Finite Element Basics</a> pages
...
 
-Gmsh is an open-source and freely available mesh generation tool with built-in CAD (Computer-aided design) and a postprocessor. The input to Gmsh can be a simple text file that provides a description of the geometry of the finite element model. The geometry can be generated using the Gmsh graphical user interface, simple text editors such as Vi/Vim/Emacs, or using more sophisticated CAD tools, like SolidWorks or Autocad. CAD models in IGES or STEP formats can be imported by the CAD engine of Gmsh, meshed, and prepared as inputs to the MFEM examples. Here, however, we will focus on more simple examples as the focus is on the process of generating meshes suitable for MFEM and not on the actual geometry. 
+Many examples together with documentation on the input syntax can be found at the [Gmsh website](https://gmsh.info/). Users familiar with Gmsh can skip the first steps and download already prepared geometries for meshing.
```
Comment (hauten) [inline src/tutorial/meshvis.md:37]: Insert a link so users can skip the mesh generation info and go straight to the downloads (line 81)?

File: src/tutorial/meshvis.md [hunk 8]
```diff
@@ -55,47 +53,56 @@ Rc=L/10;
...
 <img style="width:60%" src="../img/gmsh03.png">
 
-The generated mesh is shown in the figure above. Careful inspection reveals that the cylindrical surface is not represented well by the linear elements. We can improve the representation by refining the mesh. The Gmsh input file can be downloaded  <a href="../cross_heat.geo">here</a>  and the mesh <a href="../cross_heat.msh">here</a>. The users are strongly encouraged to play with the mesh and to generate finer discretizations for the simulations.  
+The generated mesh is shown in the figures above. Careful inspection reveals that the cylindrical surface is not represented well by the linear elements. We can improve the representation by refining the mesh. We encourage you to play with the mesh and to generate finer discretizations for the simulations.  
+
+You can download the Gmsh input file [here](../cross_heat.geo) and the mesh [here](../cross_heat.msh). For users without access to the Gmsh GUI, a mesh can be generated in a terminal with the following command:
 
-For users without access to the graphical interface of Gmsh, a mesh can be generated in a terminal by the following command: 
 ```diff
 gmsh -3 cross_heat.geo
```
Comment (hauten) [inline src/tutorial/meshvis.md:84]: Where in the mfem directory (or outside of it) should I be when running this command?

File: src/tutorial/meshvis.md [hunk 9]
```diff
@@ -55,47 +53,56 @@ Rc=L/10;
...
+To run simulations with the generated mesh, drag-and-drop the mesh file from your computer to the AWS browser window in the MFEM `examples` directory. To run Example 1 with the newly prepared mesh, be sure you are in the `examples` directory and then run the appropriate command:
+
 ```diff
-euler@26c14060d771:~/mfem/examples$ mpirun -np 24 ./ex1p -m mesh_file.msh
+mpirun -np 24 ./ex1p -m cross_heat.geo
+mpirun -np 24 ./ex1p -m cross_heat.msh
 ``` 
-The solution of the diffusion equation for the generated mesh is shown in the following two pictures. The figures are generated with ParaView, and the process of visualization is explained at the end of this tutorial session. 
+
+The solution of the diffusion equation for the generated mesh is shown in the following two pictures. The figures are generated with ParaView, and the process of visualization is explained at the end of this tutorial session.
```
Comment (hauten) [inline src/tutorial/meshvis.md:94]: Do I need to install ParaView in order to see this viz? Or add the lines from the "Visualizing results in VisIt and ParaView" section?

File: src/tutorial/meshvis.md [hunk 10]
```diff
@@ -55,47 +53,56 @@ Rc=L/10;
...
+
 <img style="width:60%" src="../img/parav00002.png">
 
-If we want to enforce Dirichlet boundary conditions different than zero on some other surface, we must export it as a physical surface. For example, if we want to enforce value one on the other cylindrical surface, we can add the following line to the geo file: 
+If we want to enforce Dirichlet boundary conditions different than zero on some other surface, we must export it as a physical surface. For example, to enforce value one on the other cylindrical surface, add the following line to the `cross_heat.geo` file:
+
 ```diff
 Physical Surface(2) = {7};
```
Comment (hauten) [inline src/tutorial/meshvis.md:101]: Does it matter where I add this line in `cross_heat.geo`?

File: src/tutorial/meshvis.md [hunk 11]
```diff
@@ -127,8 +135,8 @@ MFEM can import meshes saved in Exodus II format generated with  <a href="https:
 
 ---
 
-### <i class="fa fa-check-square-o"></i>&nbsp; Visualizing results in VisIt and Paraview
-To save the simulation results from example 1 (ex1p.cpp) in ParaView format, add the following lines just before step 17 in the file.
+### <i class="fa fa-check-square-o"></i>&nbsp; Visualizing results in VisIt and ParaView
+To save the simulation results from the parallel version of Example 1 ([ex1p.cpp](https://github.com/mfem/mfem/blob/master/examples/ex1p.cpp)) in ParaView format, add the following lines just before step 17 in the file.
```
Comment (hauten) [inline src/tutorial/meshvis.md:139]: Should the user see the "Remember to compile each numbered example before executing its sample runs" reminder here, or are they only saving data (not visualizing it)?

File: src/tutorial/meshvis.md (hunk 7 above)
Comment (bslazarov) [reply src/tutorial/meshvis.md:37]: inserted link to the geometry

File: src/tutorial/meshvis.md [hunk 12]
```diff
@@ -30,15 +30,13 @@ and <a href="../fem"><i class="fa fa-book"></i> Finite Element Basics</a> pages
 
 ### <i class="fa fa-check-square-o"></i>&nbsp; Importing meshes from Gmsh and Cubit
```
[the remark this answers is not in the document]
Comment (bslazarov) [reply src/tutorial/meshvis.md:31]: The text is modified with additional comments - the presentation should be more clear now to newcomers. Thank you!

File: src/tutorial/fem.md (hunk 1 above)
Comment (tzanio) [reply src/tutorial/fem.md:66]: Fixed

File: src/tutorial/fem.md (hunk 2 above)
Comment (tzanio) [reply src/tutorial/fem.md:123]: Thanks!

File: src/tutorial/examples.md (hunk 4 above)
Comment (tzanio) [reply src/tutorial/examples.md]: Description of corresponding GLVis keys added.

**Review by tzanio — COMMENTED** [review]
Thank you for the careful review @hauten !

**Pull request merged**